File: algorithm/baekjoon/codeplus/first-solve/src/main/java/com/keunwon/algorithm/bruteforce/Problem15664.java
package com.keunwon.algorithm.bruteforce;

import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.Scanner;
import java.util.Set;

public class Problem15664 {
    public static final Set<String> set = new LinkedHashSet<>();

    public static int[] numbers;
    public static int[] result;
    public static boolean[] visited;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final int N = sc.nextInt();
        final int M = sc.nextInt();

        numbers = new int[N];
        visited = new boolean[N];
        result = new int[M];

        for (int i = 0; i < N; i++) {
            numbers[i] = sc.nextInt();
        }

        Arrays.sort(numbers);
        dfs(N, M, 0,0);

        set.forEach(System.out::println);
    }

    public static void dfs(int n, int m, int start, int depth) {
        if (m == depth) {
            StringBuilder sb = new StringBuilder();
            for (int i : result) {
                sb.append(i).append(" ");
            }
            set.add(sb.toString());
            return;
        }

        for (int i = start; i < n; i++) {
            if (visited[i]) { continue; }

            visited[i] = true;
            result[depth] = numbers[i];
            dfs(n, m, i, depth + 1);
            visited[i] = false;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem15664.kt
package com.keunwon.algorithm.baekjoon

/**
 * Title: N과 M (10)
 * Level: 실버-2
 **/
class Problem15664 {
    private val answer = mutableListOf<String>()

    fun solution(m: Int, numbers: IntArray): List<String> {
        numbers.sort()
        dfs(0, 0, IntArray(m), numbers)

        return answer.distinct()
    }

    private fun dfs(
        depth: Int,
        startIndex: Int,
        picks: IntArray,
        numbers: IntArray,
    ) {
        if (depth == picks.size) {
            answer.add(picks.joinToString(" "))
            return
        }

        for (i in startIndex until numbers.size) {
            picks[depth] = numbers[i]
            dfs(depth + 1, i + 1, picks, numbers)
        }
    }
}

fun main() = with(System.`in`.bufferedReader()) {
    val bw = System.out.bufferedWriter()

    val (n, m) = readLine().split(" ").map { it.toInt() }
    val numbers = readLine().split(" ").map { it.toInt() }.toIntArray()

    Problem15664().solution(m, numbers).forEach {
        bw.write(it)
        bw.newLine()
    }

    bw.flush()
    bw.close()
    close()
}
